### ISSUE TITLE
Template not applied to some documents

### ISSUE THREAD

**Issue by jimmyjones2** [elastic/elasticsearch#13362]
I've distilled this from logstash, it appears that the template is not applied to the last document in some cases, so the value of field2 (cat) doesn't get included in all, so the following query:
```json
GET /repro/_search
{
  "query": {
    "filtered": {
      "query": {
        "query_string": {
          "query": "bye cat"
        }
      }
    }
  }
}
```
returns both documents the first time, then after waiting 20s when additional fields are added to the same document and the index recreated, only one is returned.

```bash
curl -XDELETE localhost:9200/repro
curl -XDELETE localhost:9200/_template/repro
curl -XPUT localhost:9200/_template/repro -d '{"template":"repro","mappings":{"_default_":{"include_in_all":false,"_all":{"analyzer":"keyword"},"properties":{"field1":{"type":"string"},"field2":{"type":"string","include_in_all":true}}}}}'
curl -XPOST localhost:9200/_bulk --data-binary '
{"index":{"_id":null,"_index":"repro","_type":"logs"}}
{"field1":"hello","field2":"bye"}
{"index":{"id":null,"_index":"repro","_type":"logs"}}
{"field1":"dog","field2":"cat"}

'
sleep 20
curl -XDELETE localhost:9200/repro
curl -XPOST localhost:9200/_bulk --data-binary '
{"index":{"_id":null,"_index":"repro","_type":"logs"}}
{"field1":"hello","field2":"bye","@version":"1"}
{"index":{"_id":null,"_index":"repro","_type":"logs"}}
{"field1":"dog","field2":"cat","@version":"1"}

'
```

**Comment by clintongormley** [comment]
Hi @jimmyjones2 

I'm guessing you're on a version before 1.6?  There is a bug (which appears to be fixed in 1.6.0) but it is different from what you describe. It can be reproduced as follows:

    DELETE *
    DELETE /_template/repro
    
    PUT /_template/repro
    {
      "template": "repro",
      "mappings": {
        "_default_": {
          "include_in_all": false,
          "_all": {
            "analyzer": "keyword"
          },
          "properties": {
            "field1": {
              "type": "string"
            },
            "field2": {
              "type": "string",
              "include_in_all": true
            }
          }
        }
      }
    }
    
Note: Your template sets the `analyzer` on the `_all` field to `keyword`, which is definitely not what you want to do for reasons described below.

Create documents with the `@version` field:

    POST _bulk 
    {"index":{"_id":null,"_index":"repro","_type":"logs"}}
    {"field1":"hello","field2":"bye","@version": "1"}
    {"index":{"_id":null,"_index":"repro","_type":"logs"}}
    {"field1":"dog","field2":"cat","@version": "1"}
    
The mapping indicates that `field1` and `@version` have `include_in_all` set to `false`, and `field2` set to `true`, which is what you expect:

    GET _mapping
    
However, the `_all` field is not as expected:

    GET _search?size=0
    {
      "aggs": {
        "NAME": {
          "terms": {
            "field": "_all",
            "size": 10
          }
        }
      }
    }

The `_all` field includes `bye`, which is correct, but it also includes the single term `cat 1` which is incorrect.  For some reason it included the `@version` field value as well, even though the final mapping is correct.

This also demonstrates why you don't want to set the analyzer for the `_all` field to `keyword`. Field values are not added to `_all` as multiple values but as a single string joined by spaces, so the `keyword` analyzer just ends up indexing a single term for the `_all` field. On top of that, the order that fields are added to `_all` is non-deterministic.

Either way, the index template/mapping issue appears to have been fixed already so I'll close this issue

**Comment by jimmyjones2** [comment]
@clintongormley Running 1.7.1 and that shows the issue. Tried 2.0beta1 and that worked fine.

The reason I used the keyword analyzer is I've got mutliple fields containing filenames which can contain spaces and I didn't want the additional terms created that the default analyzer of _all would, but I didn't realise *all* the terms marked with include_in_all got merged - thanks for clearing that up! Was trying to workaround #13214, nevermind!
